zcash_history/src/tree.rs

```diff
@@ -22,36 +43,41 @@ pub struct Tree<V: Version> {
     // number of persistent(!) tree entries
     stored_count: u32,
 
-    root: EntryLink,
+    root: V::EntryLink,
 }
 
-impl<V: Version> Tree<V> {
+impl<V> Tree<V>
+where
+    V: Version,
+    V::EntryLink: Into<EntryLink> + From<EntryLink> + Copy,
+    V::EntryKind: From<EntryKind>,
+{
     /// Resolve link originated from this tree
-    pub fn resolve_link(&self, link: EntryLink) -> Result<IndexedNode<V>, Error> {
-        match link {
+    pub fn resolve_link(&self, link: V::EntryLink) -> Result<IndexedNode<V>, Error> {
+        match link.into() {
```

Comment:
```suggestion
        match EntryLink::from(link) {
```
This is kind of trivial, in this call site there's no ambiguity, but in general `into` is unsafe because it can cause builds to break at a distance if multiple `From` impls from different locations end up in scope. As a consequence, we prefer to use `X::from(y)` to `y.into()` in general; it makes it so that you don't have to do a complete analysis of the scope to decide whether or not using `into()` is risky in that particular context.